src/rust-crypto/PerSessionKeyBackupDownloader.ts

```diff
@@ -0,0 +1,576 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import * as RustSdkCryptoJs from "@matrix-org/matrix-sdk-crypto-wasm";
+import { OlmMachine } from "@matrix-org/matrix-sdk-crypto-wasm";
+
+import { Curve25519AuthData, KeyBackupInfo, KeyBackupSession } from "../crypto-api/keybackup";
+import { IMegolmSessionData } from "../crypto";
+import { Logger } from "../logger";
+import { ClientPrefix, IHttpOpts, MatrixError, MatrixHttpApi, Method } from "../http-api";
+import { RustBackupCryptoEventMap, RustBackupCryptoEvents, RustBackupDecryptor, RustBackupManager } from "./backup";
+import { CryptoEvent, TypedEventEmitter } from "../matrix";
+import { encodeUri, sleep } from "../utils";
+import { RustCrypto } from "./rust-crypto";
+
+/**
+ * Extract the dependence of the PerSessionKeyBackupDownloader, main reason is to make testing easier.
+ */
+export interface OnDemandBackupDelegate {
+    /** Gets the current trusted backup if any.*/
+    getActiveBackupVersion(): Promise<string | null>;
+
+    /** Gets the current cached backup decryption key if any.*/
+    getBackupDecryptionKey(): Promise<RustSdkCryptoJs.BackupKeys | null>;
+
+    /**
+     *  Performs the server request to fetch the key from backup.
+     *
+     *  @param version - The backup version to use.
+     *  @param roomId - The room id of the session.
+     *  @param sessionId - The session id of the session.
+     *
+     */
+    requestRoomKeyFromBackup(version: string, roomId: string, sessionId: string): Promise<KeyBackupSession>;
+
+    /**
+     * Imports the given keys into the crypto store.
+     * @param keys - The keys to import.
+     */
+    importRoomKeys(keys: IMegolmSessionData[]): Promise<void>;
+
+    /**
+     * Creates a backup decryptor that can decrypt the key retrieved from backup.
+     * @param key - The cached backup decryption key.
+     */
+    createBackupDecryptor(key: RustSdkCryptoJs.BackupDecryptionKey): RustBackupDecryptor;
+
+    /**
+     * Requests the current key backup version from the server.
+     */
+    requestKeyBackupVersion(): Promise<KeyBackupInfo | null>;
+
+    /**
+     * The backup downloader will listen to these events to know when to check for backup status changes in order to
+     * resume or stop querying.
+     */
+    getCryptoEventEmitter(): TypedEventEmitter<RustBackupCryptoEvents, RustBackupCryptoEventMap>;
+}
+
+/**
+ * Utility to create a delegate for the PerSessionKeyBackupDownloader that is usable by rust crypto.
+ *
+ * @param rustCrypto - The rust crypto instance.
+ * @param backupManager - The backup manager instance.
+ * @param olmMachine - The olm machine instance.
+ * @param http - The http instance.
+ */
+export function createDelegate(
+    rustCrypto: RustCrypto,
```

Comment:
I'd rather we didn't have to pass the whole of RustCrypto in here: circular relationships like this make it hard to reason about the logic flows. Could we instead:

* Call `OlmMachine.importBackedUpRoomKeys` directly instead of going via `RustCrypto.importRoomKeys`
* Expose `PerSessionKeyBackupDownloader.onBackupStatusChange` and make `RustCrypto` responsible for calling it.